File: java/client/test/java/org/apache/qpid/failover/FailoverTest.java
/*
 *
 * Copyright (c) 2006 The Apache Software Foundation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
package org.apache.qpid.failover;

import org.apache.qpid.AMQException;
import org.apache.qpid.url.URLSyntaxException;
import org.apache.qpid.client.AMQConnection;
import org.apache.qpid.client.AMQConnectionURL;
import org.apache.qpid.jms.ConnectionListener;

import javax.jms.*;
import java.util.Timer;
import java.util.TimerTask;

public class FailoverTest implements MessageListener, ConnectionListener
{
    private static final long TIMEOUT = 10 * 1000;
    private static final long INTERVAL = 500;
    private final Timer _timer = new Timer(true);
    private final Connection _connection;
    private final Session _session;
    private final MessageProducer _producer;
    private Timeout _timeout;
    private int _count;
    private Queue _tempQueue;

    FailoverTest(String connectionUrl) throws JMSException, AMQException, URLSyntaxException
    {
        this(new AMQConnection(connectionUrl));
        ((AMQConnection) _connection).setConnectionListener(this);
    }

    FailoverTest(Connection connection) throws JMSException
    {
        AMQConnection amqConnection  = (AMQConnection) connection;
        System.out.println("connection.url = " + amqConnection.toURL());
        _connection = connection;
        _session = _connection.createSession(false, Session.AUTO_ACKNOWLEDGE);
        Topic topic = _session.createTopic("topic1");
        _tempQueue = _session.createTemporaryQueue();
        _producer = _session.createProducer(topic);
        _session.createConsumer(_tempQueue).setMessageListener(this);
        //new TopicListener(_session, topic);
        new TopicListener(_connection.createSession(false, Session.AUTO_ACKNOWLEDGE), topic);
        _connection.start();

        Message msg = _session.createTextMessage("Init");
        msg.setJMSReplyTo(_tempQueue);
        send(msg);
    }

    public synchronized void onMessage(Message message)
    {
        try
        {
            //cancel timeout:
            _timeout.clear();
            new DelayedSend(_session.createTextMessage("Message" + (++_count)), INTERVAL);
        }
        catch (JMSException e)
        {
            error(e);
        }
    }

    private synchronized void send(Message msg) throws JMSException
    {
        _producer.send(msg);
        //start timeout:
        _timeout = new Timeout(TIMEOUT);
    }

    private void error(Exception e)
    {
        e.printStackTrace();
        stop();
    }

    private void stop()
    {
        System.out.println("Stopping...");
        try
        {
            _connection.close();
        }
        catch (JMSException e)
        {
            System.out.println("Failed to shutdown: " + e);
            e.printStackTrace();
        }
    }

    private void timeout()
    {
        try {
            System.out.println("timed out. Resending init message");
            Message msg = _session.createTextMessage("Init");
            msg.setJMSReplyTo(_tempQueue);
            send(msg);
        } catch (JMSException e) {
            throw new RuntimeException("Got JMSException", e);
        }
//        error(new RuntimeException("Timed out: count = " + _count));
    }

    public void bytesSent(long count)
    {
    }

    public void bytesReceived(long count)
    {
    }

    public boolean preFailover(boolean redirect)
    {
        System.out.println("preFailover(" + redirect + ") called");
        return true;
    }

    public boolean preResubscribe()
    {
        System.out.println("preResubscribe() called");
        return true;
    }

    public void failoverComplete()
    {
        System.out.println("failoverComplete() called");
    }

    private class TopicListener implements MessageListener
    {
        private final Session _session;
        private MessageProducer _producer;
        private int _received;

        TopicListener(Session session, Topic topic) throws JMSException
        {
            _session = session;
            _session.createConsumer(topic).setMessageListener(this);
        }

        public void onMessage(Message message)
        {
            try
            {
                //if(_received++ % 100 == 0)
                {
                    System.out.println("Received: " + ((TextMessage) message).getText());
                }
                if(_producer == null)
                {
                    _producer = init(message);
                }
                reply(message);
            }
            catch (JMSException e)
            {
               error(e);
            }
        }

        private void reply(Message message) throws JMSException
        {
            _producer.send(_session.createTextMessage(((TextMessage) message).getText()));
        }

        private MessageProducer init(Message message) throws JMSException
        {
            return _session.createProducer(message.getJMSReplyTo());
        }
    }

    private class Timeout extends TimerTask
    {
        private volatile boolean _cancelled;

        Timeout(long time)
        {
            _timer.schedule(this, time);
        }

        void clear()
        {
            _cancelled = true;
        }

        public void run()
        {
            if(!_cancelled)
            {
                timeout();
                System.out.println("would have timed out!");
            }
        }
    }

    private class DelayedSend extends TimerTask
    {
        private final Message _msg;

        DelayedSend(Message msg, long delay)
        {
            _msg = msg;
            _timer.schedule(this, delay);
        }

        public void run()
        {
            try
            {
                send(_msg);
            }
            catch (JMSException e)
            {
                error(e);
            }
        }
    }

    public static void main(final String[] argv) throws Exception
    {
        final String clientId = "failover" + System.currentTimeMillis();
        final String defaultUrl = "amqp://guest:guest@" + clientId + "/test" +
                    "?brokerlist='tcp://localhost:5672;tcp://localhost:5673'&failover='roundrobin'";

        System.out.println("url = [" + defaultUrl + "]");

        System.out.println("connection url = [" + new AMQConnectionURL(defaultUrl) + "]");

        final String broker = argv.length == 0? defaultUrl : argv[0];
        new FailoverTest(broker);
    }
}

File: qpid/dotnet/Qpid.Client.Tests/failover/FailoverTest.cs
/*
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 * 
 *   http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 *
 */
using System;
using System.Threading;
using log4net;
using NUnit.Framework;
using Qpid.Client.Qms;
using Qpid.Messaging;

namespace Qpid.Client.Tests.failover
{
    [TestFixture, Category("Failover")]
    public class FailoverTest : IConnectionListener
    {
        private static readonly ILog _logger = LogManager.GetLogger(typeof(FailoverTest));

        private IConnection _connection;
        private IChannel _channel;
        private IMessagePublisher _publisher;
        private int _count;

        private IMessageConsumer _consumerOfResponse;

        void DoFailoverTest(IConnectionInfo info)
        {
            DoFailoverTest(new AMQConnection(info));
        }

        void DoFailoverTest(IConnection connection)
        {
            AMQConnection amqConnection = (AMQConnection)connection;
            amqConnection.ConnectionListener = this;
            //Console.WriteLine("connection.url = " + amqConnection.ToURL());
            _connection = connection;
            _connection.ExceptionListener = new ExceptionListenerDelegate(OnConnectionException);
            _channel = _connection.CreateChannel(false, AcknowledgeMode.NoAcknowledge);

            string exchangeName = ExchangeNameDefaults.TOPIC;
            string routingKey = "topic1";

            string queueName = DeclareAndBindTemporaryQueue(exchangeName, routingKey);
            
            new MsgListener(_connection.CreateChannel(false, AcknowledgeMode.NoAcknowledge), queueName);

            IChannel channel = _channel;

            string tempQueueName = channel.GenerateUniqueName();
            channel.DeclareQueue(tempQueueName, false, true, true);
            _consumerOfResponse = channel.CreateConsumerBuilder(tempQueueName).Create();
            _consumerOfResponse.OnMessage = new MessageReceivedDelegate(OnMessage);

            _connection.Start();

            IMessage msg = _channel.CreateTextMessage("Init");
            // FIXME: Leaving ReplyToExchangeName as default (i.e. the default exchange)
            // FIXME: but the implementation might not like this as it defaults to null rather than "".
            msg.ReplyToRoutingKey = tempQueueName;
//            msg.ReplyTo = new ReplyToDestination("" /* i.e. the default exchange */, tempQueueName);
            _logger.Info(String.Format("sending msg.Text={0}", ((ITextMessage)msg).Text));

//            _publisher = _channel.CreatePublisher(exchangeName, exchangeClass, routingKey);
            _publisher = _channel.CreatePublisherBuilder()
                .WithRoutingKey(routingKey)
                .WithExchangeName(exchangeName)
                .Create();
            _publisher.Send(msg);
        }

        public string DeclareAndBindTemporaryQueue(string exchangeName, string routingKey)
        {
            string queueName = _channel.GenerateUniqueName();

            // Queue.Declare
            _channel.DeclareQueue(queueName, false, true, true);

            // Queue.Bind
            _channel.Bind(queueName, exchangeName, routingKey);
            return queueName;
        }

        private void OnConnectionException(Exception e)
        {
            _logger.Error("Connection exception occurred", e);
        }

        public void OnMessage(IMessage message)
        {
            try
            {
                _logger.Info("received message on temp queue msg.Text=" + ((ITextMessage)message).Text);
                Thread.Sleep(1000);
                _publisher.Send(_channel.CreateTextMessage("Message" + (++_count)));
            }
            catch (QpidException e)
            {
                error(e);
            }
        }

        private void error(Exception e)
        {
            _logger.Error("exception received", e);
            stop();
        }

        private void stop()
        {
            _logger.Info("Stopping...");
            try
            {
                _connection.Dispose();
            }
            catch (QpidException e)
            {
                _logger.Error("Failed to shutdown", e);
            }
        }

        public void BytesSent(long count)
        {
        }

        public void BytesReceived(long count)
        {
        }

        public bool PreFailover(bool redirect)
        {
            _logger.Info("preFailover(" + redirect + ") called");
            return true;
        }

        public bool PreResubscribe()
        {
            _logger.Info("preResubscribe() called");
            return true;
        }

        public void FailoverComplete()
        {
            _logger.Info("failoverComplete() called");
        }

        private class MsgListener
        {
            private IChannel _session;
            private IMessagePublisher _publisher;

            internal MsgListener(IChannel session, string queueName)
            {
                _session = session;
                _session.CreateConsumerBuilder(queueName).Create().OnMessage = 
                    new MessageReceivedDelegate(OnMessage);
            }

            public void OnMessage(IMessage message)
            {
                try
                {
                    _logger.Info("Received: msg.Text = " + ((ITextMessage) message).Text);
                    if(_publisher == null)
                    {
                        _publisher = init(message);
                    }
                    reply(message);
                }
                catch (QpidException e)
                {
//                   Error(e);
                    _logger.Error("yikes", e); // XXX
                }
            }

            private void reply(IMessage message)
            {
                string msg = ((ITextMessage) message).Text;
                _logger.Info("sending reply - " + msg);
                _publisher.Send(_session.CreateTextMessage(msg));
            }

            private IMessagePublisher init(IMessage message)
            {
                _logger.Info(string.Format("creating reply producer with dest = '{0}:{1}'", 
                                           message.ReplyToExchangeName, message.ReplyToRoutingKey));

                string exchangeName = message.ReplyToExchangeName;
                string routingKey = message.ReplyToRoutingKey;

                //return _channel.CreatePublisher(exchangeName, exchangeClass, routingKey);
                return _session.CreatePublisherBuilder()
                    .WithExchangeName(exchangeName)
                    .WithRoutingKey(routingKey)
                    .Create();
            }
        }

        [Test]
        public void TestFail()
        {
            Assert.Fail("Tests in this class do not pass, but hang forever, so commented out until can be fixed.");
        }

        /*[Test]
        public void TestWithBasicInfo()
        {
            Console.WriteLine("TestWithBasicInfo");
            try
            {
                QpidConnectionInfo connectionInfo = new QpidConnectionInfo();
                connectionInfo.AddBrokerInfo(new AmqBrokerInfo("amqp", "localhost", 5672, false));
                connectionInfo.AddBrokerInfo(new AmqBrokerInfo("amqp", "localhost", 5673, false));               
                DoFailoverTest(connectionInfo);
                while (true)
                {
                    Thread.Sleep(5000);
                }
            }
            catch (Exception e)
            {
                _logger.Error("Exception caught", e);
            }
        }*/

//        [Test]
//        public void TestWithUrl()
//        {
//            String clientId = "failover" + DateTime.Now.Ticks;
//            String defaultUrl = "amqp://guest:guest@" + clientId + "/test" +
//                                "?brokerlist='tcp://localhost:5672;tcp://localhost:5673'&failover='roundrobin'";
//
//            _logger.Info("url = [" + defaultUrl + "]");
//
//            //            _logger.Info("connection url = [" + new AMQConnectionURL(defaultUrl) + "]");
//
//            String broker = defaultUrl;
//            //new FailoverTest(broker);
//        }
    }
}
